File: frontend/src/types/duplicateMerge.ts
/**
 * Types for the duplicate-merge surface (PR #76).
 *
 * Mirrors ``backend/app/schemas/duplicate_merge.py``. The endpoint
 * accepts a source Repository record id and merges its artifacts
 * into the target. No raw artifact internals or storage references
 * are surfaced here; that posture is enforced server-side.
 */

import type { DuplicateContractCandidate } from "./contractIntake";

export interface DuplicateMergeRequest {
  source_contract_id: string;
  /** Optional human-readable note. Not echoed back, not persisted. */
  merge_note?: string | null;
}

export interface DuplicateMergeResponse {
  target_contract_id: string;
  source_contract_id: string;
  artifacts_moved: number;
  merged_at: string;
  merged_by_user_id: string;
  /**
   * Counts of links left attached to the source after merge.
   * This PR deliberately does not rewire workflow / request links;
   * the UI uses these to render an honest "stayed on the merged
   * record" warning.
   */
  workflow_runs_attached_to_source: number;
  requests_attached_to_source: number;
}

/**
 * Response shape of ``GET /api/contracts/{id}/duplicate-candidates``.
 * Wraps the existing ``DuplicateContractCandidate`` type used at
 * upload time so the merge UI and the upload-review surface speak
 * the same vocabulary.
 */
export interface DuplicateCandidatesResponse {
  candidates: DuplicateContractCandidate[];
}

File: backend/app/schemas/duplicate_merge.py
"""Schemas for the duplicate-merge endpoint (PR #76).

Request:
    POST /api/contracts/{target_contract_id}/merge-duplicate
    Body: ``DuplicateMergeRequest``

Response:
    200 ``DuplicateMergeResponse``

Privacy posture:

* Every model uses ``extra="forbid"`` so accidental server-side
  attributes cannot leak into the wire format.
* ``merge_note`` is accepted but its text is NOT echoed back, NOT
  persisted, and NOT recorded in the audit log — only the boolean
  "a note was present" survives. The default-no posture matches the
  rest of Whereas: operator prose only leaves the deployment when
  there is a UI to read it back safely.
"""
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class DuplicateMergeRequest(BaseModel):
    """Body of ``POST /api/contracts/{target_id}/merge-duplicate``."""

    model_config = ConfigDict(extra="forbid")

    source_contract_id: uuid.UUID
    merge_note: str | None = Field(default=None, max_length=2000)


class DuplicateMergeResponse(BaseModel):
    """Outcome of a successful merge.

    ``workflow_runs_attached_to_source`` and
    ``requests_attached_to_source`` are intentionally counts only.
    This PR does NOT rewire workflow / request links; surfacing the
    counts lets the UI render an honest "these stayed on the merged
    record" warning instead of overpromising.
    """

    model_config = ConfigDict(extra="forbid")

    target_contract_id: uuid.UUID
    source_contract_id: uuid.UUID
    artifacts_moved: int
    merged_at: datetime
    merged_by_user_id: uuid.UUID
    workflow_runs_attached_to_source: int = 0
    requests_attached_to_source: int = 0
